README.md
# BusinessDays

[![Author](http://img.shields.io/badge/author-@iansltx-blue.svg?style=flat-square)](https://twitter.com/iansltx)
[![Latest Version](https://img.shields.io/github/release/iansltx/BusinessDays.svg?style=flat-square)](https://github.com/iansltx/BusinessDays/releases)
[![Software License](https://img.shields.io/badge/license-BSD-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Build Status](https://img.shields.io/travis/iansltx/BusinessDays/master.svg?style=flat-square)](https://travis-ci.org/iansltx/BusinessDays)
[![HHVM Status](https://img.shields.io/hhvm/iansltx/business-days.svg?style=flat-square)](http://hhvm.h4cc.de/package/iansltx/business-days)
[![Coverage Status](https://img.shields.io/codeclimate/coverage/github/iansltx/BusinessDays.svg?style=flat-square)](https://scrutinizer-ci.com/g/iansltx/BusinessDays/code-structure)
[![Quality Score](https://img.shields.io/scrutinizer/g/iansltx/BusinessDays.svg?style=flat-square)](https://scrutinizer-ci.com/g/iansltx/BusinessDays)
[![Code Climate](https://img.shields.io/codeclimate/github/iansltx/BusinessDays.svg?style=flat-square)](https://codeclimate.com/github/iansltx/BusinessDays)
[![SensioLabs Insight](https://img.shields.io/sensiolabs/i/933afbcf-7a28-4da0-9998-340bc7dba67c.svg?style=flat-square)]()
[![Total Downloads](https://img.shields.io/packagist/dt/iansltx/business-days.svg?style=flat-square)](https://packagist.org/packages/iansltx/business-days)

BusinessDays is a set of tools for dealing with date calculations when certain (non-business/holiday) days don't count.

This package contains two calculators, FastForwarder and Rewinder, plus a set of filters that can be used with them.

FastForwarder is configured with a day count (N) and a set of callbacks defining what *isn't* a business day, Once
configured, provide it with a start date and it'll return a date that is the first business day at least N business
days in the future. Rewinder does the same thing, except working in the opposite temporal direction.

The start date passed into the calculators can be a DateTime or DateTimeImmutable, or a subclass thereof. The value
returned will be a clone of whatever was passed in, with its timestamp updated.

This library should conform to PSRs 1, 2 and 4, and requires PHP 5.5 or newer.

## Install

Via Composer

``` bash
$ composer require iansltx/business-days
```

If you don't want Composer, you may download the source zipball directly from GitHub and load it using a PSR-4 compliant
autoloader. If you don't have such an autoloader, require `autoload.php` to get one that works for this library.

## Usage

### Filters

Calculators (FastForwarder and Rewinder) evaluate filter functions/methods/closures to see whether a given date should
be classified as a business day or not. You can provide a closure directly, or use one of the convenience methods to
compose a filter more quickly.

The convenience methods simply wrap either functions from StaticFilter or create closures via FilterFactory. Either of
these filter utility classes may be used standalone to filter DateTime objects. Additionally, these two filter classes
provide more filters than are exposed via the convenience methods on FastForwarder/Rewinder, so check them before
rebuilding a filter definition from scratch.

Note that filters requiring numeric input have internal checks that require an explicit int cast for all parameters.
Also, while skipWhen filters are executed on-add to check argument and return types, non-closure arguments are allowed,
so you can use object methods, utility functions or global functions can be used as filters in addition to closures.

Let's set up a FastForwarder, add some filters, and do some calculations...

``` php

use iansltx\BusinessDays\FilterFactory;
use iansltx\BusinessDays\StaticFilter;

// set up the instance with a day count
$ff = new iansltx\BusinessDays\FastForwarder(10);

// add a closure-based filter
$ff->skipWhen(function (\DateTimeInterface $dt) { //
    return $dt->format('m') == 12 && $dt->format('d') == 25;
}, 'christmas');

// define a closure from the filter factory, then pass it in
$dayAfterChristmasFriday = FilterFactory::monthAndDayOnDayOfWeek(12, 26, 5);
$ff->skipWhen($dayAfterChristmasFriday, 'day_after_christmas_friday');

// convenience method; saved to filter slot 'weekend'
$ff->skipWhenWeekend();

// overwrites 'weekend' slot with an identical call
$ff->skipWhen(['iansltx\BusinessDays\StaticFilter', 'isWeekend'], 'weekend');

// use some other convenience methods, this time pulling from FilterFactory and using method chaining
$ff->skipWhenNthDayOfWeekOfMonth(3, 1, 2, 'presidents_day') // third Monday of February
   ->skipWhenNthDayOfWeekOfMonth(4, 4, 11, 'thanksgiving') // fourth Thursday of November
   ->skipWhenMonthAndDay(1, 1) // auto-named to md_1_1 since a filter name wasn't provided
   ->skipWhen([StaticFilter::class, 'isGoodFriday'], 'good_friday')
   ->skipWhen([StaticFilter::class, 'isEasterMonday'], 'easter_monday');

// calculate some dates
echo $ff->exec(new \DateTime('2015-11-20 09:00:00'))->format('Y-m-d H:i:s'); // 2015-12-07 09:00:00
echo $ff->exec(new \DateTimeImmutable('2015-02-12 09:00:00'))->format('Y-m-d H:i:s'); // 2015-02-27 09:00:00

```

__NOTE:__ Fractional days are rounded up to the next whole day amount; e.g. 2.5 days will be treated as 3.

### Exporting/Importing Filter Sets

The list of filters associated with a given calculator may be dumped as an array via `getSkipWhenFilters()`. This array
can be set as filter storage for another calculation class when that class is created, so a set of filters only needs
to be defined once.

Some notes on this functionality:

1. Any filters added after construction won't propagate to the calculation object that you got the original filter
set from.
2. Passing in a set of filters at construction time will not run argument and return type tests on filters contained
therein.

With that, let's copy our filters to a new Rewinder, which as its name suggests goes in the opposite direction, and
calculate a few more dates.

``` php

// you could also create a FastForwarder with a negative day count with the same effect
$rw = new iansltx\BusinessDays\Rewinder(2, $ff->getSkipWhenFilters());

echo $rw->exec(new \DateTime('2015-02-10 09:00:00'))->format('Y-m-d H:i:s'); // 2015-02-06 09:00:00
echo $rw->exec(new \DateTime('2015-02-18 09:00:00'))->format('Y-m-d H:i:s'); // 2015-02-13 09:00:00
```

### For More Info

For more information on filter arguments etc., take a look at the source. All methods and classes have docblocks. The
callable-based syntax of skipWhen() allows for arbitrarily complex definitions of whether a date should be skipped, as
can be seen in more complex filters like isEasterMonday.

## Testing

``` bash
$ composer test
```

`humbug.json` is included if you want to do mutation testing with [Humbug](https://github.com/padraic/humbug).
Currently, not all mutations are caught; PRs are welcome to help rectify this issue. `composer test` only runs PHPUnit
tests.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details. Additional static filters/filter factories are welcome!

## Security

If you discover any security related issues, please email iansltx@gmail.com instead of using the issue tracker.

## Credits

- [Ian Littman](https://github.com/iansltx)
- [All Contributors](../../contributors)

## License

This library is BSD 2-clause licensed. Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All Notable changes to `iansltx\BusinessDays` will be documented in this file

## 1.4.0 (2015-02-21)

- Added support for negative day input on-construct (FastForwarder(-2) == Rewinder(2))
- Made constructor visibility public; supports same arguments as createWithDays(); useful for factories/DI containers
- Moved createWithDays() static method out of SkipWhenTrait

## 1.3.0 (2015-02-18)

- Added bulk filter import/export to FastForwarder
- Added Rewinder for iterating backward from, rather than forward from, a given date
- Added monthAndDaysOnDayOfWeek filter for, e.g., December 26 as long as it's a Friday
- Moved non-algorithmic functionality into SkipWhenTrait for use by other potential classes
- Fixed exec() return type when argument subclasses \DateTime or \DateTimeImmutable

## 1.2.0 (2015-02-16)

- Added isGoodFriday static filter
- Fixed #3 (Easter calculation fails on HHVM) by implementing Easter calc in PHP
  (StaticFilter::getEasterDateTimeForYear, which returns a DateTimeImmutable)

## 1.1.0 (2015-02-16)

- Added standalone autoloader
- Added isEasterMonday static filter

## 1.0.0 (2015-02-15)

Initial release, with FastForwarder plus a few FilterFactory and StaticFilter methods


tests/FilterTest.php
<?php

namespace iansltx\BusinessDays\Test;

use iansltx\BusinessDays\StaticFilter;
use iansltx\BusinessDays\FilterFactory;

class FilterTest extends \PHPUnit_Framework_TestCase
{
    public function testWeekendTrue()
    {
        $this->assertTrue(StaticFilter::isWeekend(new \DateTime('next Sunday')));
    }

    public function testWeekendFalse()
    {
        $this->assertFalse(StaticFilter::isWeekend(new \DateTime('last Tuesday')));
    }

    public function testMonthDayNonInteger()
    {
        $this->setExpectedException('\InvalidArgumentException');
        FilterFactory::monthAndDay(1, '1');
    }

    public function testMonthDayInvalidMonth()
    {
        $this->setExpectedException('\OutOfBoundsException');
        FilterFactory::monthAndDay(13, 1);
    }

    public function testMonthDayInvalidDay()
    {
        $this->setExpectedException('\OutOfBoundsException');
        FilterFactory::monthAndDay(12, 32);
    }

    public function testMonthDayInvalidDayForMonth()
    {
        $this->setExpectedException('\OutOfBoundsException');
        FilterFactory::monthAndDay(2, 30);
    }

    public function testMonthDayFalse()
    {
        $isChristmas = FilterFactory::monthAndDay(12, 25);
        $this->assertFalse($isChristmas(new \DateTime('December 26')));
    }

    public function testMonthDayTrue()
    {
        $isChristmas = FilterFactory::monthAndDay(12, 25);
        $this->assertTrue($isChristmas(new \DateTime('December 25')));
    }

    public function testMonthDayDayOfWeekNonInteger()
    {
        $this->setExpectedException('\InvalidArgumentException');
        FilterFactory::monthAndDayOnDayOfWeek(1, 1, '0');
    }

    public function testMonthDayDayOfWeekInvalidMonth()
    {
        $this->setExpectedException('\OutOfBoundsException');
        FilterFactory::monthAndDayOnDayOfWeek(13, 1, 0);
    }

    public function testMonthDayDayOfWeekInvalidDay()
    {
        $this->setExpectedException('\OutOfBoundsException');
        FilterFactory::monthAndDayOnDayOfWeek(12, 32, 0);
    }

    public function testMonthDayDayOfWeekInvalidDayForMonth()
    {
        $this->setExpectedException('\OutOfBoundsException');
        FilterFactory::monthAndDayOnDayOfWeek(2, 30, 0);
    }

    public function testMonthDayDayOfWeekInvalidDayOfWeek()
    {
        $this->setExpectedException('\OutOfBoundsException');
        FilterFactory::monthAndDayOnDayOfWeek(2, 28, 7);
    }

    public function testMonthDayDayOfWeekWrongDayOfWeek()
    {
        $isChristmas = FilterFactory::monthAndDayOnDayOfWeek(12, 26, 5);
        $this->assertFalse($isChristmas(new \DateTime('December 26 2015')));
    }

    public function testMonthDayDayOfWeekWrongMonth()
    {
        $isChristmas = FilterFactory::monthAndDayOnDayOfWeek(12, 26, 5);
        $this->assertFalse($isChristmas(new \DateTime('September 26 2014')));
    }

    public function testMonthDayDayOfWeekWrongDay()
    {
        $isChristmas = FilterFactory::monthAndDayOnDayOfWeek(12, 26, 5);
        $this->assertFalse($isChristmas(new \DateTime('December 25 2015')));
    }

    public function testMonthDayDayOfWeekTrue()
    {
        $isChristmas = FilterFactory::monthAndDayOnDayOfWeek(12, 26, 5);
        $this->assertTrue($isChristmas(new \DateTime('December 26 2014')));
    }

    public function testNthDayOfWeekOfMonthNonInteger()
    {
        $this->setExpectedException('\InvalidArgumentException');
        FilterFactory::nthDayOfWeekOfMonth(1, 0, '1');
    }

    public function testNthDayOfWeekOfMonthInvalidN()
    {
        $this->setExpectedException('\OutOfBoundsException');
        FilterFactory::nthDayOfWeekOfMonth(6, 0, 1);
    }

    public function testNthDayOfWeekOfMonthInvalidDayOfWeek()
    {
        $this->setExpectedException('\OutOfBoundsException');
        FilterFactory::nthDayOfWeekOfMonth(1, 7, 1);
    }

    public function testNthDayOfWeekOfMonthInvalidMonth()
    {
        $this->setExpectedException('\OutOfBoundsException');
        FilterFactory::nthDayOfWeekOfMonth(1, 1, 13);
    }

    public function testNthDayOfWeekOfMonthFalseMonth()
    {
        $isThanksgiving = FilterFactory::nthDayOfWeekOfMonth(4, 4, 11);
        $this->assertFalse($isThanksgiving(new \DateTime('October 26 2015')));
    }

    public function testNthDayOfWeekOfMonthFalseWeek()
    {
        $isThanksgiving = FilterFactory::nthDayOfWeekOfMonth(4, 4, 11);
        $this->assertFalse($isThanksgiving(new \DateTime('November 19 2015')));
    }

    public function testNthDayOfWeekOfMonthFalseDayOfWeek()
    {
        $isThanksgiving = FilterFactory::nthDayOfWeekOfMonth(4, 4, 11);
        $this->assertFalse($isThanksgiving(new \DateTime('November 25 2015')));
    }

    public function testNthDayOfWeekOfMonthTrue()
    {
        $isThanksgiving = FilterFactory::nthDayOfWeekOfMonth(4, 4, 11);
        $this->assertTrue($isThanksgiving(new \DateTime('November 26 2015')));
    }

    public function testEasterMondayWrongYear()
    {
        $this->assertFalse(StaticFilter::isEasterMonday(new \DateTime('April 6 2014')));
    }

    public function testEasterMondayWrongMonth()
    {
        $this->assertFalse(StaticFilter::isEasterMonday(new \DateTime('March 6 2015')));
    }

    public function testEasterMondayWrongDay()
    {
        $this->assertFalse(StaticFilter::isEasterMonday(new \DateTime('April 5 2015')));
    }

    public function testEasterMondayMatch()
    {
        $this->assertTrue(StaticFilter::isEasterMonday(new \DateTime('April 6 2015 23:59:59')));
    }

    public function testGoodFridayWrongYear()
    {
        $this->assertFalse(StaticFilter::isGoodFriday(new \DateTime('April 3 2014')));
    }

    public function testGoodFridayWrongMonth()
    {
        $this->assertFalse(StaticFilter::isGoodFriday(new \DateTime('March 3 2015')));
    }

    public function testGoodFridayWrongDay()
    {
        $this->assertFalse(StaticFilter::isGoodFriday(new \DateTime('April 4 2015')));
    }

    public function testGoodFridayMatch()
    {
        $this->assertTrue(StaticFilter::isGoodFriday(new \DateTime('April 3 2015 23:59:59')));
    }
}


tests/DateTimeExtended.php
<?php

namespace iansltx\BusinessDays\Test;

use DateTimeZone;

class DateTimeExtended extends \DateTime
{
    protected $somethingElse;

    public function __construct($time = 'now', DateTimeZone $timezone = null)
    {
        parent::__construct($time, $timezone);
        $this->somethingElse = rand();
    }

    public function getSomethingElse() {
        return $this->somethingElse;
    }
}

tests/FastForwarderTest.php
<?php

namespace iansltx\BusinessDays\Test;

use iansltx\BusinessDays\FastForwarder;
use iansltx\BusinessDays\Rewinder;
use DateTimeImmutable as Immutable;
use DateTime as Mutable;
use DateInterval as Interval;
use iansltx\BusinessDays\StaticFilter;

class FastForwarderTest extends \PHPUnit_Framework_TestCase
{
    public function testTimeZoneSafety()
    {
        $ff = FastForwarder::createWithDays(10);
        $startDt = new Immutable();
        $expectedDt = $startDt->add(new Interval('P10D'));

        $this->assertEquals($expectedDt, $ff->exec($startDt));
    }

    public function testImmutable()
    {
        $ff = $this->addFilterSet1(FastForwarder::createWithDays(10));

        $endDt = $ff->exec(new Immutable('2015-02-12 09:00:00'));

        $this->assertEquals('2015-02-27 09:00:00', $endDt->format('Y-m-d H:i:s'));
        $this->assertInstanceOf('\DateTimeImmutable', $endDt);
    }

    public function testExtendedClass()
    {
        $ff = $this->addFilterSet1(FastForwarder::createWithDays(10));

        $extendedDt = new DateTimeExtended('2015-02-12 09:00:00');
        $somethingElse = $extendedDt->getSomethingElse();

        /** @var DateTimeExtended $endDt */
        $endDt = $ff->exec($extendedDt);

        $this->assertEquals('2015-02-27 09:00:00', $endDt->format('Y-m-d H:i:s'));
        $this->assertInstanceOf(DateTimeExtended::class, $endDt);
        $this->assertEquals($somethingElse, $endDt->getSomethingElse());
    }

    public function testMutable()
    {
        $ff = $this->addFilterSet1(FastForwarder::createWithDays(10));

        $startDt = new Mutable('2015-11-20 09:00:00');
        $endDt = $ff->exec($startDt);

        $this->assertEquals('2015-12-07 09:00:00', $endDt->format('Y-m-d H:i:s'));
        $this->assertInstanceOf('\DateTime', $endDt);
        $this->assertEquals('2015-11-20 09:00:00', $startDt->format('Y-m-d H:i:s'));
    }

    public function testRewindOverWeekend()
    {
        $rw = $this->addFilterSet1(Rewinder::createWithDays(2));

        $endDt = $rw->exec(new Mutable('2015-02-10 09:00:00'));

        $this->assertEquals('2015-02-06 09:00:00', $endDt->format('Y-m-d H:i:s'));
        $this->assertInstanceOf('\DateTime', $endDt);
    }

    public function testRewindOverPresidentsDay()
    {
        $rw = $this->addFilterSet1(Rewinder::createWithDays(2));

        $endDt = $rw->exec(new Mutable('2015-02-18 09:00:00'));

        $this->assertEquals('2015-02-13 09:00:00', $endDt->format('Y-m-d H:i:s'));
        $this->assertInstanceOf('\DateTime', $endDt);
    }

    public function testNegativeDayCount()
    {
        $rw = $this->addFilterSet1(new FastForwarder(-2));

        $endDt = $rw->exec(new Mutable('2015-02-18 09:00:00'));

        $this->assertEquals('2015-02-13 09:00:00', $endDt->format('Y-m-d H:i:s'));
        $this->assertInstanceOf('\DateTime', $endDt);
    }

    public function testRewindOverBusinessDays()
    {
        $rw = $this->addFilterSet1(Rewinder::createWithDays(2));

        $endDt = $rw->exec(new Mutable('2015-02-19 09:00:00'));

        $this->assertEquals('2015-02-17 09:00:00', $endDt->format('Y-m-d H:i:s'));
        $this->assertInstanceOf('\DateTime', $endDt);
    }

    public function testPartialDays()
    {
        $rw = $this->addFilterSet1(Rewinder::createWithDays(2.5));

        $endDt = $rw->exec(new Mutable('2015-02-19 09:00:00'));

        // rounds up to 3, so skips Presidents Day and weekend
        $this->assertEquals('2015-02-13 09:00:00', $endDt->format('Y-m-d H:i:s'));
        $this->assertInstanceOf('\DateTime', $endDt);
    }

    public function testSkipWhenException()
    {
        $ff = FastForwarder::createWithDays(1);

        $this->setExpectedException('InvalidArgumentException');
        $ff->skipWhen(function (\DateTimeInterface $dt) {return $dt->format('m');}, 'strict_typehints_ftw');
    }

    public function testFilterImportExport()
    {
        $ff = $this->addFilterSet1(FastForwarder::createWithDays(10));
        $filters = $ff->getSkipWhenFilters();

        $this->assertCount(7, $filters); // takes into account weekend being overwritten
        $this->assertArrayHasKey('md_1_1', $filters); // automated filter name generation
        $this->assertArrayHasKey('christmas', $filters); // manual filter name generation
        $this->assertArrayNotHasKey(0, $filters); // not a numerically indexed array

        $ff2 = FastForwarder::createWithDays(10, $filters);

        $endDt = $ff2->exec(new Mutable('2015-11-20 09:00:00'));

        $this->assertEquals('2015-12-07 09:00:00', $endDt->format('Y-m-d H:i:s'));
        $this->assertInstanceOf('\DateTime', $endDt);
    }

    protected function addFilterSet1(FastForwarder $ff)
    {
        return $ff->skipWhenWeekend()
            ->skipWhen([StaticFilter::class, 'isWeekend'], 'weekend') // overwrites convenience method above
            ->skipWhenNthDayOfWeekOfMonth(3, 1, 2, 'presidents_day')
            ->skipWhenNthDayOfWeekOfMonth(4, 4, 11, 'thanksgiving')
            ->skipWhenMonthAndDay(1, 1) // test auto-naming
            ->skipWhen(['iansltx\BusinessDays\StaticFilter', 'isEasterMonday'], 'easter_monday')
            ->skipWhen(['iansltx\BusinessDays\StaticFilter', 'isGoodFriday'], 'good_friday')
            ->skipWhen(function (\DateTimeInterface $dt) { // test providing a callable directly
                    return $dt->format('m') == 12 && $dt->format('d') == 25;
                }, 'christmas');
    }
}


autoload.php
<?php // PSR-4 autoloader for internal classes

spl_autoload_register(function ($class) {
    $map = [
        'iansltx\\BusinessDays\\Test\\' => __DIR__ . DIRECTORY_SEPARATOR . 'tests' . DIRECTORY_SEPARATOR,
        'iansltx\\BusinessDays\\' => __DIR__ . DIRECTORY_SEPARATOR . 'src' . DIRECTORY_SEPARATOR
    ];

    foreach ($map as $prefix => $baseDir) {
        // does the class use the namespace prefix?
        $len = strlen($prefix);
        if (strncmp($prefix, $class, $len) !== 0) {
            // no, move to the next registered autoloader
            continue;
        }

        // get the relative class name
        $relative_class = substr($class, $len);

        // replace the namespace prefix with the base directory, replace namespace
        // separators with directory separators in the relative class name, append
        // with .php
        $file = $baseDir . str_replace('\\', DIRECTORY_SEPARATOR, $relative_class) . '.php';

        // if the file exists, require it
        if (file_exists($file)) {
            require $file;
        }
    }
});


src/Rewinder.php
<?php

namespace iansltx\BusinessDays;

/**
 * Class Rewinder
 *
 * Provides an easy way to calculate dates that are a specified number of
 * "business bays" in the past relative to a supplied date. This is achieved
 * by first adding a series of filter callbacks that define what is NOT a
 * business day.
 *
 * Calculates a date X days before $start_date, where X was supplied in
 * static::createWithDays(); if the end date would land on a non-business
 * day, the last business day before that date is returned.
 *
 * A negative day count may be entered into the constructor to count in the
 * opposite direction (days in the future vs. in the past). Or just use
 * FastForwarder.
 *
 * @package iansltx\BusinessDays
 */
class Rewinder extends FastForwarder
{
    public function __construct($num_days, array $skip_when = [])
    {
        parent::__construct($num_days * -1, $skip_when);
    }
}


src/FastForwarder.php
<?php

namespace iansltx\BusinessDays;

use DateTimeInterface;
use DateTime;
use DateInterval;

/**
 * Class FastForwarder
 *
 * Provides an easy way to calculate dates that are a specified number of
 * "business bays" in the future relative to a supplied date. This is achieved
 * by first adding a series of filter callbacks that define what is NOT a
 * business day.
 *
 * Calculates a date X days after $start_date, where X was supplied in
 * static::createWithDays(); if the end date would land on a non-business
 * day, the first business day after that date is returned.
 *
 * A negative day count may be entered into the constructor to count in the
 * opposite direction (days in the past vs. in the future). Or just use
 * Rewinder.
 *
 * @package iansltx\BusinessDays
 */
class FastForwarder
{
    use SkipWhenTrait;

    protected $numDays = 0;

    /** @var DateInterval */
    protected $interval;

    /**
     * Static factory method for backward compatibility; see constructor for more details
     *
     * @param int|float $num_days
     * @param array $skip_when
     * @return static
     */
    public static function createWithDays($num_days, array $skip_when = [])
    {
        return new static($num_days, $skip_when);
    }

    /**
     * Creates an instance with a defined number of business days
     *
     * @param int|float $num_days the number of business days from the supplied date to
     *  the calculated date (non-integer values will be rounded up)
     * @param array $skip_when pre-defined filters to add to skipWhen without testing
     */
    public function __construct($num_days, array $skip_when = [])
    {
        $this->numDays = ceil(abs($num_days));
        $this->interval = new DateInterval('P1D');
        // invert iterator direction if negative day count
        $this->interval->invert = (int) ($num_days < 0);
        $this->replaceSkipWhen($skip_when);
    }

    /**
     * Iterates through dates in steps of $this->interval until $this->numDays
     * is zero. Then, if the current day is not a business day, iterate until
     * the current day is a business day. Return the result in the same format
     * as was supplied.
     *
     * @param DateTimeInterface $start_date
     * @return DateTimeInterface clone of $start_date with a modified timestamp
     */
    public function exec(DateTimeInterface $start_date)
    {
        $startTz = $start_date->getTimezone();
        $newDt = (new DateTime($start_date->format('c')))->setTimeZone($startTz);

        $daysLeft = $this->numDays;

        while ($daysLeft > 0) {
            if ($this->getSkippedBy($newDt) === false) {
                --$daysLeft;
            }

            $newDt->add($this->interval);
        }

        do { // if end date falls on a skipped day, keep skipping until a valid day is found
            $skipFilterName = $this->getSkippedBy($newDt);
            if ($skipFilterName !== false) {
                $newDt->add($this->interval);
            }
        } while ($skipFilterName !== false);

        /** @var DateTime|\DateTimeImmutable $clonedDt */
        $clonedDt = clone $start_date;
        return $clonedDt->setTimestamp($newDt->getTimestamp());
    }
}


src/FilterFactory.php
<?php

namespace iansltx\BusinessDays;

/**
 * Class FilterFactory
 *
 * Builds filter functions based on dynamic input
 *
 * @package iansltx\BusinessDays
 */
class FilterFactory
{
    /**
     * Returns a filter that returns true if the day and month match,
     * false otherwise; the month is represented as an integer (January
     * is 1, December is 12)
     *
     * @param int $month
     * @param int $day
     * @return callable
     * @throws \OutOfBoundsException|\InvalidArgumentException
     */
    public static function monthAndDay($month, $day)
    {
        static::checkMonthAndDayArgs($month, $day);

        return function(\DateTimeInterface $dt) use ($month, $day) {
            return (int) $dt->format('m') === $month && (int) $dt->format('d') === $day;
        };
    }

    /**
     * Returns a filter that returns true if the day, month and day of week
     * match, false otherwise; the month and day of week are represented as
     * integers (January is 1, December is 12, Sunday is 0, Saturday is 6)
     *
     * @param int $month
     * @param int $day
     * @param int $day_of_week
     * @return callable
     * @throws \OutOfBoundsException|\InvalidArgumentException
     */
    public static function monthAndDayOnDayOfWeek($month, $day, $day_of_week)
    {
        static::checkMonthAndDayArgs($month, $day);

        if (!is_int($day_of_week)) {
            throw new \InvalidArgumentException('$day_of_week must be an integer');
        }

        if ($day_of_week < 0 || $day_of_week > 6) {
            throw new \OutOfBoundsException('$day_of_week must be 0-6');
        }

        return function(\DateTimeInterface $dt) use ($month, $day, $day_of_week) {
            return (int) $dt->format('m') === $month && (int) $dt->format('d') === $day &&
                    (int) $dt->format('w') === $day_of_week;
        };
    }

    /**
     * Returns a filter that returns true if the date is the Nth (e.g. 4rd)
     * day (e.g. Thursday, as an integer where Sunday is 0 and Saturday is 6)
     * of a given month (as an integer where January is 1 and December is 12)
     *
     * @param int $n
     * @param int $day_of_week
     * @param int $month
     * @return callable
     * @throws \OutOfBoundsException|\InvalidArgumentException
     */
    public static function nthDayOfWeekOfMonth($n, $day_of_week, $month)
    {
        static::checkNthDayArgs($n, $day_of_week, $month);

        $lowerBound = ($n - 1) * 7;
        $upperBound = $n * 7;

        return function(\DateTimeInterface $dt) use ($month, $day_of_week, $lowerBound, $upperBound) {
            if ((int) $dt->format('m') !== $month) {
                return false;
            }

            return ((int) $dt->format('w')) === $day_of_week &&
                    $dt->format('d') > $lowerBound && $dt->format('d') <= $upperBound;
        };
    }

    /**
     * Does type and bounds checks for nThDayOfWeekOfMonth()
     *
     * @param $n
     * @param $day_of_week
     * @param $month
     */
    protected static function checkNthDayArgs($n, $day_of_week, $month)
    {
        if (!is_int($n) || !is_int($day_of_week) || !is_int($month)) {
            throw new \InvalidArgumentException('$n, $day_of_week and $month must be integers');
        }

        if ($n < 1 || $n > 5) {
            throw new \OutOfBoundsException('$n must be 1-5');
        }

        if ($day_of_week < 0 || $day_of_week > 6) {
            throw new \OutOfBoundsException('$day_of_week must be 0-6');
        }

        if ($month < 1 || $month > 12) {
            throw new \OutOfBoundsException('$month must be 1-12');
        }
    }

    /**
     * Does type and bounds checks for monthAndDay()
     *
     * @param $month
     * @param $day
     */
    protected static function checkMonthAndDayArgs($month, $day)
    {
        if (!is_int($month) || !is_int($day)) {
            throw new \InvalidArgumentException('$month and $day must be integers');
        }

        if ($month < 1 || $month > 12) {
            throw new \OutOfBoundsException('$month must be 1-12');
        }

        if ($day < 1 || $day > 31) {
            throw new \OutOfBoundsException('$day must be a valid day of the month');
        }

        if ((in_array($month, [4, 6, 9, 11]) && $day > 30) || ((int) $month === 2 && $day > 29)) {
            throw new \OutOfBoundsException('Day ' . $day . ' does not exist in month ' . $month);
        }
    }
}


src/PeriodIterator.php
<?php

namespace iansltx\BusinessDays;

use DateTimeInterface;
use DateTime;
use DateTimeImmutable;
use DateInterval;

abstract class PeriodIterator
{
    use SkipWhenTrait;

    protected function __construct()
    {
    }

    /**
     * @param DateTimeInterface $start_date
     * @param DateInterval $step
     * @param bool $is_reversed
     * @return DateTime|DateTimeImmutable
     */
    protected function iterate(DateTimeInterface $start_date, DateInterval $step, $is_reversed = false)
    {
        $startTz = $start_date->getTimezone();
        $newDt = (new DateTime($start_date->format('c')))->setTimeZone($startTz);

        $daysLeft = $this->numDays;

        while ($daysLeft > 0) {
            if ($this->getSkippedBy($newDt) === false) {
                --$daysLeft;
            }

            !$is_reversed ? $newDt->add($step) : $newDt->sub($step);
        }

        do { // if end date falls on a skipped day, keep skipping until a valid day is found
            $skipFilterName = $this->getSkippedBy($newDt);
            if ($skipFilterName !== false) {
                !$is_reversed ? $newDt->add($step) : $newDt->sub($step);
            }
        } while ($skipFilterName !== false);

        return $start_date instanceof DateTime ? $newDt->setTimezone($startTz) :
            new DateTimeImmutable($newDt->format('Y-m-d H:i:s'), $startTz);
    }

    /**
     * @param DateTimeInterface $start_date
     * @return DateTimeInterface
     */
    abstract public function exec(DateTimeInterface $start_date);
}


src/StaticFilter.php
<?php

namespace iansltx\BusinessDays;

/**
 * Class StaticFilter
 *
 * A library of static filter functions that can be used directly
 *
 * @package iansltx\BusinessDays
 */
class StaticFilter
{
    protected static $easterByYear = []; // cache of Easter DateTimeImmutable objects, indexed by year

    /**
     * Returns true if $dt is on a Saturday or Sunday, false otherwise
     *
     * @param \DateTimeInterface $dt
     * @return bool
     */
    public static function isWeekend(\DateTimeInterface $dt)
    {
        return in_array($dt->format('w'), [0, 6]);
    }

    /**
     * Returns true if $dt is the Monday after Western Easter, false otherwise
     *
     * @param \DateTimeInterface $dt
     * @return bool
     */
    public static function isEasterMonday(\DateTimeInterface $dt)
    {
        $easterMonday = static::getEasterDateTimeForYear($dt->format('Y'))->add(new \DateInterval('P1D'));
        return $easterMonday->format('m') === $dt->format('m') && $easterMonday->format('d') === $dt->format('d');
    }

    /**
     * Returns true if $dt is the Friday before Western Easter, false otherwise
     *
     * @param \DateTimeInterface $dt
     * @return bool
     */
    public static function isGoodFriday(\DateTimeInterface $dt)
    {
        $goodFriday = static::getEasterDateTimeForYear($dt->format('Y'))->sub(new \DateInterval('P2D'));
        return $goodFriday->format('m') === $dt->format('m') && $goodFriday->format('d') === $dt->format('d');
    }

    /**
     * Gets an immutable DateTime object set to midnight UTC of Western Easter
     * for a given year. Built because HHVM doesn't include easter_date() or
     * easter_day(). Credit to https://www.drupal.org/node/1180480 for the
     * original implementation.
     *
     * To get a runtime-agnostic result comparable to PHP4+'s easter_date(),
     * call getTimestamp() on the returned value of this function.
     *
     * @param int $year
     * @return \DateTimeImmutable
     */
    public static function getEasterDateTimeForYear($year) {
        if (!isset(static::$easterByYear[$year])) { // build cached value if it doesn't exist
            $century = intval($year / 100);
            $centuryOverFour = intval($century / 4);
            $centuryMod4 = $century % 4;

            $twoDigitYear = $year % 100;
            $twoDigitYearOver4 = intval($twoDigitYear / 4);
            $twoDigitYearMod4 = $twoDigitYear % 4;

            $yearMod19 = $year % 19;

            $interF = intval(($century + 8) / 25);
            $interG = intval(($century - $interF + 1) / 3);
            $interH = (19 * $yearMod19 + $century - $centuryOverFour - $interG + 15) % 30;
            $interL = (32 + 2 * $centuryMod4 + 2 * $twoDigitYearOver4 - $interH - $twoDigitYearMod4) % 7;
            $interM = intval(($yearMod19 + 11 * $interH + 22 * $interL) / 451);
            $interP = ($interH + $interL - 7 * $interM + 114) % 31;

            $monthNumber = intval(($interH + $interL - 7 * $interM + 114) / 31);
            $day = $interP + 1;

            static::$easterByYear[$year] = \DateTimeImmutable::createFromFormat(
                   'Y-m-d', $year . '-' . $monthNumber . '-' . $day, new \DateTimeZone('UTC'));
        }

        return static::$easterByYear[$year]; // return cached value
    }
}
